Mathlib/Logic/Function/CompTypeclasses.lean

```diff
@@ -0,0 +1,65 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import Mathlib.Init.Function
+
+/-!
+# Propositional typeclasses on several maps
+
+This file contains typeclasses that are used in the definition of
+equivariant maps in the spirit what was initially developed
+by Frédéric Dupuis and Heather Macbeth for linear maps.
+
+* `CompTriple φ ψ χ`, which expresses that `ψ.comp φ = χ`
+* `CompTriple.IsId φ`, which expresses that `φ = id`
+
+TODO :
+* align with RingHomCompTriple
+
+-/
+
+section CompTriple
+
+/-- Class of composing triples -/
+class CompTriple  {M N P : Type*} (φ : M → N) (ψ : N → P) (χ : outParam (M → P)) : Prop where
+  /-- The maps form a commuting triangle -/
+  comp_eq : ψ.comp φ = χ
+
+attribute [simp] CompTriple.comp_eq
+
+namespace CompTriple
+
+/-- Class of Id maps -/
+class IsId {M : Type*} (σ : M → M) : Prop where
+  eq_id : σ = id
+
+instance {M : Type*} : IsId (@id M) where
+  eq_id := rfl
+
+instance instComp_id {N P : Type*} {φ : N → N} [IsId φ] {ψ : N → P} :
+    CompTriple φ ψ ψ where
+  comp_eq := by simp only [IsId.eq_id, Function.comp_id]
+
+instance instId_comp {M N : Type*} {φ : M → N} {ψ : N → N} [IsId ψ] :
+    CompTriple φ ψ φ where
+  comp_eq := by simp only [IsId.eq_id, Function.id_comp]
+
+instance (priority := 10) instComp {M N P : Type*}
```

Comment:
I would like to know if this instance is crucial for your applications. Because it always applies, it seems a bit dangerous to me.
I'm interested in your thoughts about this.

Reply:
Initially, there were no instances. I had two versions of composition, one with a comp parameter as an instance, and the other one where it had to be supplied explicitly. 
I was worried to add instances, because Heather and Frédéric seem to be worried in their files about semilinear maps, but @j-loreaux suggested we could have everything as instances. And I tried, and it worked, so in the recent version, everything is an instance, and everything compiles without ever having to fill in the instances.